// css-literal-loader/lib/plugin.js
"use strict";

exports.__esModule = true;
exports.default = plugin;

var _commonTags = require("common-tags");

var _fsExtra = require("fs-extra");

var _camelCase = _interopRequireDefault(require("lodash/camelCase"));

var _get = _interopRequireDefault(require("lodash/get"));

var _kebabCase = _interopRequireDefault(require("lodash/kebabCase"));

var _path = require("path");

var t = _interopRequireWildcard(require("@babel/types"));

var _template = _interopRequireDefault(require("@babel/template"));

var _generator = _interopRequireDefault(require("@babel/generator"));

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const buildImport = (0, _template.default)('require(FILENAME);');
const buildComponent = (0, _template.default)(`styled(TAGNAME, OPTIONS, DISPLAYNAME, IMPORT, KEBABNAME, CAMELNAME)`);
const STYLES = Symbol('CSSLiteralLoader');

function getIdentifier(path) {
  const parent = path.findParent(p => p.isVariableDeclarator());
  return parent && t.isIdentifier(parent.node.id) ? parent.node.id.name : '';
}

function wrapInClass(className, value) {
  const imports = [];
  let match;
  const matcher = /@import.*?(?:$|;)/g; // eslint-disable-next-line no-cond-assign

  while (match = matcher.exec(value)) {
    imports.push(match[0]);
  }

  value = value.replace(matcher, '');
  let val = `${className} {\n${value}\n}`;
  if (imports.length) val = `${imports.join('\n')}\n${val}`;
  return val;
}

function createFileName(hostFile, {
  extension = '.css'
}, id) {
  const base = (0, _path.basename)(hostFile, (0, _path.extname)(hostFile));
  return (0, _path.join)((0, _path.dirname)(hostFile), `${base}-${id}${extension}`);
}

function isCssTag(path, tagName, allowGlobal = false) {
  return path.get('tag.name').node === tagName && (path.get('tag').referencesImport('css-literal-loader/styled') || allowGlobal && path.scope.hasGlobal(tagName));
}

const isStyledTag = (path, styledTag, allowGlobal) => {
  const {
    node
  } = path.get('tag');
  return t.isCallExpression(node) && node.callee.name === styledTag && (allowGlobal || path.get('tag.callee').referencesImport('css-literal-loader/styled'));
};

const isStyledTagShorthand = (path, styledTag, allowGlobal) => {
  console.log();
  return t.isMemberExpression(path.get('tag').node) && t.isIdentifier(path.get('tag.property').node) && path.get('tag.object').node.name === styledTag && (allowGlobal || path.get('tag.object').referencesImport('css-literal-loader/styled'));
};

function plugin() {
  function evaluate(path) {
    const {
      confident,
      value
    } = path.evaluate();

    if (!confident) {
      throw path.buildCodeFrameError('Could not evaluate css. Inline css must be statically analyzable');
    }

    return value;
  }

  function createStyleNode(path, {
    opts,
    file
  }, identifier) {
    const {
      start,
      end
    } = path.node;
    const style = {
      start,
      end
    };
    const getFileName = opts.getFileName || createFileName;
    const hostFile = file.opts.filename;
    style.path = getFileName(hostFile, opts, identifier);
    let filename = (0, _path.relative)((0, _path.dirname)(hostFile), style.path);

    if (!filename.startsWith('.')) {
      filename = `./${filename}`;
    }

    style.filename = filename;
    style.identifier = identifier;
    return style;
  }

  function buildStyleRequire(path, state) {
    const {
      styles
    } = state.file.get(STYLES);
    const quasiPath = path.get('quasi');
    const style = createStyleNode(path, state, getIdentifier(path));
    style.value = evaluate(quasiPath);
    style.code = `require('${style.filename}')`;
    styles.add(style);
    return buildImport({
      FILENAME: t.StringLiteral(style.filename)
    }); // eslint-disable-line new-cap
  }

  function buildStyledComponent(path, tagName, options, state) {
    const cssState = state.file.get(STYLES);
    const displayName = getIdentifier(path) || tagName.value;
    const style = createStyleNode(path, state, displayName);
    style.tagName = t.isStringLiteral(tagName) ? `"${tagName.value}"` : tagName;
    const kebabName = (0, _kebabCase.default)(displayName);
    style.value = wrapInClass(`.${kebabName}`, evaluate(path.get('quasi')));
    const runtimeNode = buildComponent({
      TAGNAME: tagName,
      OPTIONS: options || t.NullLiteral(),
      DISPLAYNAME: t.StringLiteral(displayName),
      IMPORT: buildImport({
        FILENAME: t.StringLiteral(style.filename)
      }).expression,
      KEBABNAME: t.StringLiteral(kebabName),
      CAMELNAME: t.StringLiteral((0, _camelCase.default)(kebabName))
    });
    if (state.opts.generateInterpolations) style.code = (0, _generator.default)(runtimeNode).code;
    cssState.styles.add(style);
    return runtimeNode;
  }

  return {
    pre(file) {
      if (!file.has(STYLES)) {
        file.set(STYLES, {
          id: 0,
          styles: new Set()
        });
      }
    },

    post(file) {
      const {
        opts
      } = this;
      let {
        styles
      } = file.get(STYLES);
      styles = Array.from(styles.values());
      file.metadata['css-literal-loader'] = {
        styles
      };

      if (opts.writeFiles !== false) {
        styles.forEach(({
          path,
          value
        }) => {
          (0, _fsExtra.outputFileSync)(path, (0, _commonTags.stripIndent)([value]));
        });
      }
    },

    visitor: {
      TaggedTemplateExpression(path, state) {
        const {
          tagName = 'css',
          allowGlobal = true,
          styledTag = 'styled'
        } = state.opts;

        if (isStyledTag(path, styledTag, allowGlobal)) {
          const componentType = (0, _get.default)(path.get('tag'), 'node.arguments[0]');
          const options = (0, _get.default)(path.get('tag'), 'node.arguments[1]');
          path.replaceWith(buildStyledComponent(path, componentType, options, state));
          path.addComment('leading', '#__PURE__'); // styled.button` ... `
        } else if (isStyledTagShorthand(path, styledTag, allowGlobal)) {
          const componentType = t.StringLiteral(path.get('tag.property').node.name);
          path.replaceWith(buildStyledComponent(path, componentType, null, state));
          path.addComment('leading', '#__PURE__'); // lone css`` tag
        } else if (isCssTag(path, tagName, allowGlobal)) {
          path.replaceWith(buildStyleRequire(path, state));
          path.addComment('leading', '#__PURE__');
        }
      }

    }
  };
}
